src/hotspot/share/classfile/fieldLayoutBuilder.cpp

```diff
@@ -657,7 +657,8 @@ void FieldLayoutBuilder::regular_field_sorting() {
           //too_volatile_to_flatten = false; //FIXME
           // volatile fields are currently never inlined, this could change in the future
         }
-        if (!(too_big_to_flatten | too_atomic_to_flatten | too_volatile_to_flatten) || fieldinfo.access_flags().is_final()) {
+        if (!too_big_to_flatten &&
+            (!(too_atomic_to_flatten | too_volatile_to_flatten) || fieldinfo.access_flags().is_final())) {
```

Comment:
Hi @XiaohongGong , specifications (https://openjdk.org/projects/valhalla/design-notes/state-of-valhalla/03-vm-model#flattenable-contexts) also talks about flattening final L descriptor fields of value objects.

Reply:
Thanks for pointing out this! I got through of the specifications just now.
```
Final fields containing value objects, under both L and Q descriptors, can be routinely flattened. Mutable fields containing value objects under Q descriptors can be routinely flattened; mutable fields containing value objects under L descriptors must provide atomic loads and stores, which requires either atomic instructions or an indirection.
```
Per my understanding, this means:
1. The final fields can be flattened for all L/Q descriptors
2. The non-final fields can be flattened for Q descriptors
3. The non-final fields cannot be flattened for L descriptor value objects due to the atomic feature.

```
For value objects with many fields, the JVM may decide to fall back to buffering its value in the heap, which is the traditional indirected representation.
```
And the above means size of the flattened fields should not exceed some value. Right? 

Please correct me if any misunderstanding! Thanks!

Reply:
Yes, this is correct.

Reply:
The state of Valhalla document is not a specification, but a document about the general design of the Valhalla features in the VM. Currently, fields with L-descriptors are never flattened even the final ones. There are some non-trivial problems to be solved regarding their status before initialization and until we have a good solution to those problems, the safest option is to not flattened them.

Reply:
Thanks @fparain

Reply:
> The state of Valhalla document is not a specification, but a document about the general design of the Valhalla features in the VM. Currently, fields with L-descriptors are never flattened even the final ones. There are some non-trivial problems to be solved regarding their status before initialization and until we have a good solution to those problems, the safest option is to not flattened them.

null initialized L descriptor final fields may be tricky to handle, being pessimist looks appropriate in this case.

Reply:
Sounds reasonable to me! It also reminds me that the value object JEP [1] said a value object can only be scalarized when it is on stack or passed through a method call in C2 compiler. 

So maybe changing to `!(too_big_to_flatten | too_atomic_to_flatten | too_volatile_to_flatten)` is safer for current status. WDYT @fparain  @jatin-bhateja ?
 
[1] https://openjdk.org/jeps/8277163

Reply:
Looks fine to me, the case where we are making this check is meant for T_PRIMITIVE_OBJECTS (Q descriptor value object), thus as per above understanding (1 and 2), final constraint can be removed.  Value object anyways carry an L descriptor (T_OBJECT) and hence are currently not being handled.